_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+|Specialized to terraform
+
+|VCS integration
+|Yes, with plugins
+|Yes
+|Yes
+|Yes
+|Yes
+
+|Provides static IP addresses for IP whitelisting
+|Yes
+|No
+|Yes (self-hosted)
+|Yes
+|Yes (TFE)
+
+|Update commit statuses
+|Yes
+|Yes
+|Yes
+|Yes
+|Yes
+
+|Annotate pull requests
+|Requires custom scripting
+|Requires custom scripting
+|Requires custom scripting
+|Yes
+|Yes
+
+|Supports multiple infrastructure tools
+|Yes
+|Yes
+|Yes
+|Limited (terraform only; additional binaries can be installed, but can not be called directly)
+|No (terraform only; TFE can support additional binaries, but TFC does not)
+
+|===
+
+
+[[production_grade_design]]
+== Production-grade design
+
+With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]
+
+
+[[use_generic_cicd_platforms_as_a_workflow_engine_but_run_infrastructure_deployments_from_within_your_account]]
+=== Use generic CI/CD platforms as a workflow engine but run infrastructure deployments from within your account
+
+Given the limitations and tradeoffs of the various platforms we covered in <<cicd_platforms>>, we don't recommend
+relying on a single platform for implementing the entire workflow. Instead, we recommend a hybrid solution that takes
+advantage of the strengths of each platform, and cover the weaknesses. The design looks as follows:
+
+- Use any generic CI/CD server you wish to integrate with your VCS repositories.
+- Deploy a self-hosted deployment server within your AWS account that cannot be accessed by anyone but a few admins plus
+  a limited trigger we'll define below.
+- Define your CI workflows so that the CI/CD server triggers deployments against the deploy server.
+
+This design implements separation of the concerns so that we take full advantage of the strengths of each platform,
+while covering the weaknesses: relying on the CI/CD platforms to manage the workflow/pipeline, but have it trigger
+infrastructure deployments on self-hosted systems that are more locked down.
+
+We don't want to give the CI/CD servers permissions to deploy and manage arbitrary infrastructure. CI/CD servers are
+typically not secure enough to handle sensitive information, and you don't want a server that is used for executing
+arbitrary code and regularly used (and written to) by your entire dev team to have admin permissions.
+
+Instead, we delegate this responsibility to an isolated, closed off system in the AWS account that only expose a limited
+set of actions that can be triggered. That way, if anyone gets access to your CI server, they can at most kick off
+builds on existing code, but they don't get arbitrary admin access.
+
+
+[[options_for_deploy_server]]
+=== Options for deploy server
```

Comment:
Observation: This section directly contrasts Gruntwork ECS Deploy Runner Stack with Terraform Enterprise, and paints Gruntwork's solution in a better light. Them are fightin' words.

Reply:
That was not the intention, so highlighted the strengths of TFE better: 0093729

What do you think of the updated wording?